Index: support/rrsync
===================================================================
--- support/rrsync
+++ support/rrsync	2008-12-01 16:52:49.000000000 +0100
@@ -51,7 +51,7 @@
 # To disable a short-named option, add its letter to this string:
 our $short_disabled = 's';
 
-our $short_no_arg = 'ACDEHIKLORSWXbcdgklmnoprstuvxz'; # DO NOT REMOVE ANY
+our $short_no_arg = 'ACDEHIKLORSWXbcdgiklmnoprstuvxz'; # DO NOT REMOVE ANY
 our $short_with_num = 'B'; # DO NOT REMOVE ANY
 
 # To disable a long-named option, change its value to a -1.  The values mean:
