--- rsync-3.1.1pre1/Makefile.in.whole_program~	2013-06-16 01:40:10.000000000 +0200
+++ rsync-3.1.1pre1/Makefile.in	2014-04-03 17:05:19.291126189 +0200
@@ -44,6 +44,7 @@ DAEMON_OBJ = params.o loadparm.o clients
 popt_OBJS=popt/findme.o  popt/popt.o  popt/poptconfig.o \
 	popt/popthelp.o popt/poptparse.o
 OBJS=$(OBJS1) $(OBJS2) $(OBJS3) $(DAEMON_OBJ) $(LIBOBJ) @BUILD_ZLIB@ @BUILD_POPT@
+SRCS=$(patsubst %.o,%.c,$(OBJS))
 
 TLS_OBJ = tls.o syscall.o lib/compat.o lib/snprintf.o lib/permstring.o lib/sysxattrs.o @BUILD_POPT@
 
@@ -89,8 +90,13 @@ install-all: install install-ssl-client
 install-strip:
 	$(MAKE) INSTALL_STRIP='-s' install
 
+ifndef @WHOLEPROGRAM@
+rsync$(EXEEXT): rounding.h $(SRCS)
+	$(CC) $(CFLAGS) -I. -I$(srcdir) $(CPPFLAGS) $(LDFLAGS) -fwhole-program -flto -o $@ $^ $(LIBS)
+else
 rsync$(EXEEXT): $(OBJS)
 	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OBJS) $(LIBS)
+endif
 
 $(OBJS): $(HEADERS)
 $(CHECK_OBJS): $(HEADERS)
--- rsync-3.1.1pre1/configure.ac.whole_program~	2014-01-26 18:32:43.000000000 +0100
+++ rsync-3.1.1pre1/configure.ac	2014-04-03 17:02:37.834272923 +0200
@@ -795,6 +795,23 @@ else
     AC_MSG_RESULT(no)
 fi
 
+AC_ARG_ENABLE(wholeprogram,
+	AC_HELP_STRING([--enable-wholeprogram], [compile all source at once using -fwhole-program]))
+
+if test x"$enable_wholeprogram" != x"no"; then
+    define_wholeprogram=yes
+fi
+
+AC_MSG_CHECKING([whether to compile everything at once with -fwhole-program])
+if test x"$enable_wholeprogram" = x"yes"; then
+    AC_MSG_RESULT(yes)
+    CFLAGS="$CFLAGS"
+    WHOLEPROGRAM=1
+    AC_SUBST(WHOLEPROGRAM)
+else
+    AC_MSG_RESULT(no)
+fi
+
 AC_CACHE_CHECK([for unsigned char],rsync_cv_SIGNED_CHAR_OK,[
 AC_TRY_COMPILE([],[signed char *s = ""],
 rsync_cv_SIGNED_CHAR_OK=yes,rsync_cv_SIGNED_CHAR_OK=no)])
--- rsync-3.1.1pre1/lib/inet_ntop.c.whole_program~	2006-10-16 20:11:24.000000000 +0200
+++ rsync-3.1.1pre1/lib/inet_ntop.c	2014-04-03 17:02:37.834272923 +0200
@@ -26,6 +26,7 @@
  * sizeof(int) < 4.  sizeof(int) > 4 is fine; all the world's not a VAX.
  */
 
+#ifndef __socklen_t_defined
 static const char *inet_ntop4(const unsigned char *src, char *dst,
 			      size_t size);
 
@@ -184,3 +185,4 @@ inet_ntop6(const unsigned char *src, cha
 	return (dst);
 }
 #endif /* AF_INET6 */
+#endif
